general/cuda.hpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_CUDA_HPP
+#define MFEM_CUDA_HPP
+
+#include <cstddef>
+
+#ifdef MFEM_USE_CUDA
+#include <cuda.h>
+#endif
+
+namespace mfem
+{
+
+#ifdef MFEM_USE_CUDA
+#define MFEM_HOST_DEVICE __host__ __device__
+inline void CuCheck(const unsigned int c)
+{
+   MFEM_ASSERT(c == cudaSuccess, cudaGetErrorString(cudaGetLastError()));
+}
+template<typename T> MFEM_HOST_DEVICE
+inline T AtomicAdd(T* address, T val)
+{
+   return atomicAdd(address, val);
```

Comment:
atomicAdd for doubles is not supported in CUDA 8 and below.  Do you have an implementation of that flowting around somewhere?

Reply:
Yes, I'll add the fallback

Reply:
done with the `okina-atomic` merge and `MFEM_CUDA_ARCH`

Reply:
@jakubcerveny, can you try `make cuda MFEM_CUDA_ARCH=sm_52`?

@camierjs, on my Linux box (with `sm_30`) `./ex1 -d cuda` gives me an error (`PtrKnown bytes==0`)

Reply:
This works now, thank you.